lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -526,7 +530,39 @@ void StackAllocationPromoter::fixPhiPredBlock(BlockSet &phiBlocks,
   deleter.forceDelete(ti);
 }
 
-void StackAllocationPromoter::fixBranchesAndUses(BlockSet &phiBlocks) {
+bool StackAllocationPromoter::isUnnecessaryPhi(
+    SILPhiArgument *phi, const BlockSetVector &phiBlocks,
+    SmallPtrSetImpl<SILPhiArgument *> &visitedPhis) {
+  // A phi is unnecessary if it has no uses
+  if (phi->use_empty())
+    return true;
+
+  // If already visited, don't explore further.
+  if (visitedPhis.contains(phi))
+    return true;
```

Comment:
`return true` looks wrong here. E.g. 
```
bb1(%phi1):
  br bb2(%phi1)
bb2(%phi2): // not dead!
  // use %phi2
```
Let's say `isUnnecessaryPhi` is called for `%phi1` first - and returns false (which is correct). When `isUnnecessaryPhi` is then called for `%phi2`, which is already in `visitedPhis`, it will return `true`.

I don't think you can solve the dead-phi-removal problem by walking the def-use graph in "forward" direction.
Instead, you can first mark all used phi arguments and then propagate those flags backward. Like it's done in DeadCodeElimination.

Reply:
I clear the `visitedPhis` map in the main loop before successive calls. It's not actually a cache. Going to look at DCE now.

Reply:
ok, I didn't see that. But then the problem is that the algorithm is quadratic - in case of long phi-chains.

BTW, as you can have only one phi argument per block, you could also use a BasicBlockSet: `visitedBlock` instead of `visitedPhis`. With that you don't need to keep the set around for multiple `isUnnecessaryPhi` calls for performance reasons. You can just declare it inside `isUnnecessaryPhi`

Reply:
I am following the approach of backward propagating like DCE now. Thanks for pointing me to that. I did leave the new `livePhis` as a `SmallPtrSet` for readability reasons.